R/create_credentials.R

```diff
@@ -0,0 +1,233 @@
+#' package first being available on the system. We can use
+#' `install.packages("keyring")` to install
+#'
+#' @param id An option to specify an identifying label for the keyname. If no
+#'   `id` is provided, an autogenerated keyname will be constructed in the
+#'   following way: `blastula-v1-<host_name>`.
+#' @inheritParams create_smtp_creds_file
+#' @examples
+#' \dontrun{
+#' # Store SMTP crendentials using the
+#' # system's secure key-value store to
+#' # make it much easier to send email
+#' # out through Gmail with `smtp_send()`;
+#' # provide the `id` of "gmail_creds"
+#' create_smtp_creds_key(
+#'   id = "gmail_creds",
+#'   provider = "gmail",
+#'   user = "user_name@gmail.com",
+#'   )
+#' }
+#' @export
+create_smtp_creds_key <- function(id = NULL,
+                                  provider = NULL,
+                                  user,
+                                  host = NULL,
+                                  port = NULL,
+                                  use_ssl = NULL,
+                                  sender_name = NULL) {
+
+  # Creating credential on the system-wide key-value
+  # store requires the installation of the keyring package
+  if (!requireNamespace("keyring", quietly = TRUE)) {
+
+    stop("The `keyring` package is required for using the ",
+         "`create_smtp_creds_key()` function",
+         call. = FALSE)
+
+  } else {
+
+    # Determine whether {keyring} can be used
+    is_keyring_capable()
+
+    # Create a credentials list from the function inputs
+    credentials_list <-
+      create_credentials_list(
+        provider = provider,
+        user = user,
+        sender_name = sender_name,
+        host = host,
+        port = port,
+        use_ssl = use_ssl
+      )
+
+    # Create the `service_name` value for `keyring`
+    service_name <- normalize_name(name = id, host = credentials_list$host)
+
+    # Create a plaintext JSON string for the credentials
+    serialized <- JSONify_credentials(credentials_list)
+
+    # Set the key in the system's default keyring
+    keyring::key_set_with_value(
+      service = service_name,
+      user = user,
+      password = serialized
+    )
+
+    # Issue a message stating that the file has been created
+    message("The system key store has been updated with the (`", service_name, "`) key.")
+  }
+}
+
+#' Create a credentials list object
+#'
+#' @noRd
+create_credentials_list <- function(provider,
+                                    user,
+                                    sender_name,
+                                    host,
+                                    port,
+                                    use_ssl) {
+
+  # Prompt for the password using `getPass::getPass()`
+  password <- getPass::getPass("Enter the SMTP server password: ")
+
+  # If a `provider` name is given, extract values for `host`,
+  # `port`, `use_ssl`, `use_tls`, and `authenticate`
+  if (!is.null(provider) &&
```

Comment:
Throw an error if `provider` isn't found in `smtp_settings` (i.e. not supported). The error should include the list of supported providers.

Reply:
Now checks that `provider` value is supported. Throws with informative stop message now.